src/transformers/models/mt5/modeling_mt5.py

```diff
@@ -997,18 +997,13 @@ def forward(
             if not self.is_decoder:
                 raise ValueError(f"`use_cache` can only be set to `True` if {self} is used as a decoder")
 
-        if attention_mask is None:
-            attention_mask = torch.ones(batch_size, mask_seq_length, device=inputs_embeds.device)
-        if self.is_decoder and encoder_attention_mask is None and encoder_hidden_states is not None:
-            encoder_seq_length = encoder_hidden_states.shape[1]
-            encoder_attention_mask = torch.ones(
-                batch_size, encoder_seq_length, device=inputs_embeds.device, dtype=torch.long
-            )
-
         # initialize past_key_values with `None` if past does not exist
         if past_key_values is None:
             past_key_values = [None] * len(self.block)
 
+        if attention_mask is None:
+            attention_mask = torch.ones(batch_size, mask_seq_length, device=inputs_embeds.device, dtype=torch.long)
```

Comment:
Sorry - I wasn't clear enough on may last comment. For the creation of `attention_mask` and `encoder_attention_mask` we need both to have the exact same instantiation as before to avoid any unintended changes to current behaviour. So for `attention_mask` we don't want to have `dtype=torch.long`. 

```suggestion
            attention_mask = torch.ones(batch_size, mask_seq_length, device=inputs_embeds.device)
```

Reply:
The returned dtype from the tokenizer _should_ be `torch.int64` i.e. long. If it's float, then I'd be interested to know the checkpoint as that shouldn't be the case! 

This means of course that this logic is a bit inconsistent. However, most of the time the models are called with inputs prepared by the tokenizer, so this code path is never touched. 

Unfortunately, even though setting `dtype=torch.long` is more correct, part of managing the codebase is handling backwards compatibility. I actually don't think this change would cause issues, however it's very hard to know especially for popular models like T5 and tracking numerical issues is a pain 😅

Reply:
Fully understand and agree that backward compatibility is essential!

Even though feeding `input_ids`, `labels`, and `attention_mask` to the forward method of T5 is usually enough, sometimes it's essential to have more control. For example, to compute the likelihood of continuations, i.e. `-logp(continuation | context)` for evaluation purposes (e.g., the OpenLLM leaderboard). That's why I started looking into this code :)

The backstory is that I was trying to understand what the authors of the T-Few paper were doing. To compute `logp(continuation | context)` for T5, it is essential to be able to cache the encoder outputs (i.e., the encoded context) and pass the continuations through the model to collect the loglikelihoods. In this case, you need to pass `decoder_input_ids` yourself. tl;dr: it felt strange to me how the authors computed the `decoder_attention_mask` (see https://github.com/r-three/t-few/issues/32) and to try to understand their code I started to investigate what was done by default in T5 :)